src/endpoint/s3/s3_rest.js

```diff
@@ -267,6 +270,86 @@ function authenticate_request(req, res) {
     }
 }
 
+async function authorize_request(req) {
+    if (req.op_name === 'put_bucket') return;
+    const policy_info = req.params.bucket && await req.object_sdk.read_bucket_sdk_policy_info(req.params.bucket);
+    const s3_policy = policy_info.s3_policy;
+    const system_owner = policy_info.system_owner;
+    const bucket_owner = policy_info.bucket_owner;
+    if (s3_policy) {
+        const arn_path = _get_arn_from_req_path(req);
+        const method = _get_method_from_req(req);
+        const account = (await req.object_sdk.rpc_client.account.read_account({}));
+        // system owner can always change bucket policy
+        // bucket owner has FC ACL by default - so no need to check bucket policy
+        if (((account.email.unwrap() === system_owner.unwrap()) && req.op_name.endsWith('bucket_policy')) ||
+            account.email.unwrap() === bucket_owner.unwrap()) return;
+        if (!has_bucket_policy_permission(s3_policy, account.email, method, arn_path)) {
+            throw new S3Error(S3Error.AccessDenied);
+        }
+    }
+}
+
+function _get_method_from_req(req) {
+    if (req.query && req.query.versionId && s3_utils.OP_NAME_TO_ACTION[req.op_name].versioned) {
```

Comment:
If the OP does not exist on the list you will get an exception. 
Maybe it is better to check prior and have a dedicated error for such a case.